ivy_tests/test_ivy/test_misc/test_tt_tensor.py

```diff
@@ -0,0 +1,142 @@
+        ivy.TTTensor.pad_tt_rank(tt, n_padding=n_pad, pad_boundaries=False)
+    )
+    rec = tt.to_tensor()
+    rec_padded = padded_tt.to_tensor()
+
+    np.testing.assert_array_almost_equal(rec, rec_padded, decimal=4)
+    np.testing.assert_(padded_tt.rank == (1, *[i + n_pad for i in rank[1:-1]], 1))
+
+
+# TODO: Uncomment once ivy.tensor_train is implemented
+# @pytest.mark.parametrize(
+#     "shape, rank",
+#     [((3, 4, 5, 6, 2, 10), 10)],
+# )
+# def test_tt_to_tensor_random(shape, rank):
+#     tensor = ivy.random_uniform(shape)
+#     tensor_shape = tensor.shape
+
+#     factors = ivy.tensor_train(tensor, rank)
+#     reconstructed_tensor = ivy.TTTensor.tt_to_tensor(factors)
+#     np.testing.assert_(ivy.shape(reconstructed_tensor) == tensor_shape)
+
+#     D = len(factors)
+#     for k in range(D):
+#         (r_prev, _, r_k) = factors[k].shape
+#         assert r_prev <= rank, "TT rank with index " + str(k) + "exceeds rank"
+#         assert r_k <= rank, "TT rank with index " + str(k + 1) + "exceeds rank"
+
+
+@pytest.mark.parametrize(
+    "shape, rank",
+    [((4, 5, 4, 8, 5), (1, 3, 2, 2, 4, 1))],
+)
+def test_tt_n_param(shape, rank):
+    factors = ivy.random_tt(shape, rank)
+    true_n_param = ivy.sum([ivy.prod(f.shape) for f in factors])
+    n_param = ivy.TTTensor._tt_n_param(shape, rank)
+    np.testing.assert_equal(n_param, true_n_param)
+
+
+@pytest.mark.parametrize(
+    "n1, n2, n3, shape1, shape2, shape3",
+    [(3, 4, 2, (1, 3, 2), (2, 4, 2), (2, 2, 1))],
+)
+def test_tt_to_tensor(n1, n2, n3, shape1, shape2, shape3):
+    tensor = ivy.zeros((n1, n2, n3))
+
+    for i in range(n1):
+        for j in range(n2):
+            for k in range(n3):
+                tensor[i][j][k] = (i + 1) + (j + 1) + (k + 1)
+
+    tensor = ivy.array(tensor)
+
+    factors = [None] * 3
+
+    factors[0] = ivy.zeros(shape1)
+    factors[1] = ivy.zeros(shape2)
+    factors[2] = ivy.zeros(shape3)
+
+    for i in range(3):
+        for j in range(4):
+            for k in range(2):
+                factors[0][0][i][0] = i + 1
+                factors[0][0][i][1] = 1
+
+                factors[1][0][j][0] = 1
+                factors[1][0][j][1] = 0
+                factors[1][1][j][0] = j + 1
+                factors[1][1][j][1] = 1
+
+                factors[2][0][k][0] = 1
+                factors[2][1][k][0] = k + 1
+
+    factors = [ivy.array(f) for f in factors]
+
+    np.testing.assert_array_almost_equal(tensor, ivy.TTTensor.tt_to_tensor(factors))
+
+
+@pytest.mark.parametrize(
+    "coef",
+    [((0.2))],
+)
+def test_validate_tt_rank(coef):
+    tensor_shape = tuple(ivy.random.randint(5, 10, shape=(4,)))
+    n_param_tensor = ivy.prod(tensor_shape)
+
+    # TODO: This test fails even for the native implementation
+    # rank = ivy.TTTensor.validate_tt_rank(tensor_shape, coef, rounding="floor")
+    # n_param = ivy.TTTensor._tt_n_param(tensor_shape, rank)
+    # np.testing.assert_(n_param >= n_param_tensor * coef)
```

Comment:
I'm not sure if I follow this, If our implementation matches exactly with TensorLy, this should ideally pass as well. What's the error you're getting on running this?

Reply:
I'm getting an assertion error. This test fails for TensorLy as well, e.g:
```
import ivy
import numpy as np
import tensorly as tl

coef = 0.2

tensor_shape = tuple(np.random.randint(5, 10, size=4))
n_param_tensor = np.prod(tensor_shape)

# ivy
rank = ivy.TTTensor.validate_tt_rank(tensor_shape, coef, rounding="floor")
n_param = ivy.TTTensor._tt_n_param(tensor_shape, rank)
np.testing.assert_(n_param >= n_param_tensor * coef)

# tensorly
rank = tl.tt_tensor.validate_tt_rank(tensor_shape, coef, rounding="floor")
n_param = tl.tt_tensor._tt_n_param(tensor_shape, rank)
np.testing.assert_(n_param >= n_param_tensor * coef)
```

Reply:
Only `ceil` mode is passing